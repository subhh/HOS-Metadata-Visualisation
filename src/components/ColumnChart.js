import React from 'react'
import { entries, values, max, keys } from 'lodash'
import { scaleLinear, scaleTime, timeYear, extent } from 'd3'
import {
  ChartContainer,
  Title,
  ColumnsContainer,
  Group,
  Line,
  AxisLabel,
  Column,
  Tick,
} from './ChartAtoms'
import ChartLegend from './ChartLegend'

const AxisLine = ({ value, width, label, padding }) => (
  <Group y={value}>
    <Line ml={padding} width={width} bg="greyLight" />
    <AxisLabel textAlign="right">{label}</AxisLabel>
  </Group>
)

/**
 * Column chart component
 */
const ColumnChart = ({
  data,
  legend,
  height = 250,
  pt = 0, // padding top
  pb = 32, // padding bottom
  pl = 48, // padding left
  pr = 32, // padding right
  bw = 8, // bar width
  bs = 4, // bar spacing
  labelFreq = 5,
  title = 'title',
}) => {
  const k = keys(data)
  const v = values(data)
  const entr = entries(data)

  // y scale
  const yScale = scaleLinear()
    .domain([0, max(v) || 0])
    .range([height - pb, pt])
  const baseY = yScale(0)

  // x scale
  const count = v.length
  const width = count * bw + bs * (count - 1) + pl + pr
  const xScale = scaleTime()
    .domain(extent(k).map(v => new Date(v)))
    .range([pl, width])
    .nice()

  const ticks = xScale.ticks(timeYear.every(3))
  const formatTick = xScale.tickFormat('%Y')

  return (
    <ChartContainer>
      <Title>{title}</Title>
      <ColumnsContainer width={width} height={height} mb={4}>
        {yScale.ticks(4).map(t => (
          <AxisLine
            key={`yAxis-${t}`}
            label={t}
            value={yScale(t)}
            padding={pl}
            width={width - pl}
          />
        ))}
        {ticks.map((t, i) => (
          <Group key={`tick-${i}`} x={xScale(t)} y={0}>
            <Tick y={baseY + bs * 2} />
            <AxisLabel y={baseY + bs * 6} textAlign="center">
              {formatTick(t)}
            </AxisLabel>
          </Group>
        ))}
        {entr.map(([key, value], i) => {
          const x = xScale(new Date(key))
          const y = yScale(value)
          const bh = baseY - y

          return (
            <Group key={`column-${x}-${y}`} x={x} y={0}>
              <Column
                x={-bw / 2}
                y={y}
                width={bw}
                height={bh}
                highlight={i >= entr.length - 3}
              />
            </Group>
          )
        })}
      </ColumnsContainer>
      {legend && <ChartLegend reverse data={legend} direction="row" />}
    </ChartContainer>
  )
}

export default ColumnChart
