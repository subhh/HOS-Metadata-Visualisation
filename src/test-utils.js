import React, { useReducer } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { MemoryRouter } from 'react-router-dom'
import { StateContext, DispatchContext, reducer, initialState } from './state'
import theme from './theme'

const AllTheProviders = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  return (
    <ThemeProvider theme={theme}>
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={dispatch}>
          {children}
        </DispatchContext.Provider>
      </StateContext.Provider>
    </ThemeProvider>
  )
}

const WithRouter = ({ children }) => (
  <MemoryRouter>
    <AllTheProviders>{children}</AllTheProviders>
  </MemoryRouter>
)

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export const customRenderWithRouter = (ui, options) =>
  render(ui, {
    wrapper: WithRouter,
    ...options,
  })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }

export const mockAsync = data => {
  jest.spyOn(global, 'fetch').mockImplementation(url => {
    return Promise.resolve({
      json: () => Promise.resolve(data),
    })
  })
}

export const restoreAsyncMock = () => {
  global.fetch.mockRestore()
}
