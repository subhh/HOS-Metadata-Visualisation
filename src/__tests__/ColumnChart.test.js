import React from 'react'
import { render } from 'test-utils'
import ColumnChart from 'components/ColumnChart'

const data = { var1: 10, var2: 30 }

it('<ColumnChart /> without data should match snapshot', () => {
  const { container } = render(<ColumnChart />)
  expect(container).toMatchSnapshot()
})

it('<ColumnChart /> with data should match snapshot', () => {
  const { container } = render(<ColumnChart data={data} />)
  expect(container).toMatchSnapshot()
})
