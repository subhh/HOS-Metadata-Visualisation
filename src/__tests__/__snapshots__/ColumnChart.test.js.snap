// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColumnChart /> with data should match snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa sc-htoDjs bZvMhP"
  >
    <h2
      class="sc-bdVaJa sc-htpNat sc-bxivhb sc-dnqmqq fCKnkB"
      color="blueDark"
      font-size="2"
      font-weight="500"
    >
      title
    </h2>
    <div
      class="sc-bdVaJa sc-kAzzGY kVBcOy"
      height="250"
      width="100"
    >
      <div
        class="sc-bdVaJa sc-chPdSV izGVhe"
        y="218"
      >
        <div
          class="sc-bdVaJa sc-kgoBCf liHRjq"
          width="52"
        />
        <div
          class="sc-bdVaJa sc-htpNat sc-kpOJdX gfIBzX"
          color="grey"
          font-size="0"
          width="40"
        >
          0
        </div>
      </div>
      <div
        class="sc-bdVaJa sc-chPdSV vSWnj"
        y="145.33333333333334"
      >
        <div
          class="sc-bdVaJa sc-kgoBCf liHRjq"
          width="52"
        />
        <div
          class="sc-bdVaJa sc-htpNat sc-kpOJdX gfIBzX"
          color="grey"
          font-size="0"
          width="40"
        >
          10
        </div>
      </div>
      <div
        class="sc-bdVaJa sc-chPdSV ifnPcl"
        y="72.66666666666669"
      >
        <div
          class="sc-bdVaJa sc-kgoBCf liHRjq"
          width="52"
        />
        <div
          class="sc-bdVaJa sc-htpNat sc-kpOJdX gfIBzX"
          color="grey"
          font-size="0"
          width="40"
        >
          20
        </div>
      </div>
      <div
        class="sc-bdVaJa sc-chPdSV pRZwl"
        y="0"
      >
        <div
          class="sc-bdVaJa sc-kgoBCf liHRjq"
          width="52"
        />
        <div
          class="sc-bdVaJa sc-htpNat sc-kpOJdX gfIBzX"
          color="grey"
          font-size="0"
          width="40"
        >
          30
        </div>
      </div>
      <div
        class="sc-bdVaJa sc-chPdSV pRZwl"
        y="0"
      >
        <div
          class="sc-bdVaJa sc-kGXeez gVfvLo"
          height="72.66666666666666"
          width="8"
          x="-4"
          y="145.33333333333334"
        />
      </div>
      <div
        class="sc-bdVaJa sc-chPdSV pRZwl"
        y="0"
      >
        <div
          class="sc-bdVaJa sc-kGXeez kNIGie"
          height="218"
          width="8"
          x="-4"
          y="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ColumnChart /> without data should match snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa sc-htoDjs bZvMhP"
  >
    <h2
      class="sc-bdVaJa sc-htpNat sc-bxivhb sc-dnqmqq fCKnkB"
      color="blueDark"
      font-size="2"
      font-weight="500"
    >
      title
    </h2>
    <div
      class="sc-bdVaJa sc-kAzzGY bQitgL"
      height="250"
      width="76"
    >
      <div
        class="sc-bdVaJa sc-chPdSV kTYTZz"
        y="109"
      >
        <div
          class="sc-bdVaJa sc-kgoBCf loeOHy"
          width="28"
        />
        <div
          class="sc-bdVaJa sc-htpNat sc-kpOJdX gfIBzX"
          color="grey"
          font-size="0"
          width="40"
        >
          0
        </div>
      </div>
    </div>
  </div>
</div>
`;
